ci/svd2rust-regress/src/main.rs

```diff
@@ -0,0 +1,161 @@
+    #[structopt(short = "c", long = "chip")]
+    chip: Option<String>,
+
+    /// Filter by manufacturer, case sensitive, may be combined with other filters
+    #[structopt(short = "m", long = "manufacturer")]
+    mfgr: Option<String>,
+
+    /// Filter by architecture, case sensitive, may be combined with other filters
+    /// Options are: "CortexM", "RiscV", and "Msp430"
+    #[structopt(short = "a", long = "architecture")]
+    arch: Option<String>,
+
+    /// Include tests expected to fail (will cause a non-zero return code)
+    #[structopt(short = "b", long = "bad-tests")]
+    bad_tests: bool,
+
+    // TODO: Specify smaller subset of tests? Maybe with tags?
+    // TODO: Early fail
+    // TODO: Compile svd2rust?
+}
+
+/// Validate any assumptions made by this program
+fn validate_tests(tests: &[&tests::TestCase]) {
+    use std::collections::HashSet;
+
+    let mut fail = false;
+
+    // CONDITION 1: All mfgr+chip names must be unique
+    let mut uniq = HashSet::new();
+    for t in tests {
+        let name = t.name();
+        if !uniq.insert(name.clone()) {
+            fail = true;
+            eprintln!("{} is not unique!", name);
+        }
+    }
+
+    if fail {
+        panic!("Tests failed validation");
+    }
+}
+
+fn main() {
+    let opt = Opt::from_args();
+
+    // Validate all test pre-conditions
+    validate_tests(tests::TESTS);
+
+    // Determine default svd2rust path
+    let default_svd2rust_iter = ["..", "..", "..", "..", "target", "release"];
+
+    let default_svd2rust = if cfg!(windows) {
+        default_svd2rust_iter.iter().chain(["svd2rust.exe"].iter())
+    } else {
+        default_svd2rust_iter.iter().chain(["svd2rust"].iter())
+    }.collect();
+
+    let bin_path = match opt.bin_path {
+        Some(ref bp) => bp,
+        None => &default_svd2rust,
+    };
+
+    // collect enabled tests
+    let tests = tests::TESTS
+        .iter()
+        // Short test?
+        .filter(|t| t.should_run(!opt.long_test))
+        // selected architecture?
+        .filter(|t| {
+            if let Some(ref arch) = opt.arch {
+                arch == &format!("{:?}", t.arch)
+            } else {
+                true
+            }
+        })
+        // selected manufacturer?
+        .filter(|t| {
+            if let Some(ref mfgr) = opt.mfgr {
+                mfgr == &format!("{:?}", t.mfgr)
+            } else {
+                true
+            }
+        })
+        // Specify chip - note: may match multiple
+        .filter(|t| {
+            if let Some(ref chip) = opt.chip {
+                chip == t.chip
+            } else {
+                true
+            }
+        })
+        // Run failable tests?
+        .filter(|t| opt.bad_tests || t.should_pass)
+        .collect::<Vec<_>>();
+
+    let any_fails = AtomicBool::new(false);
+
+    // TODO: It would be more efficient to reuse directories, so we don't
+    // have to rebuild all the deps crates
```

Comment:
You can place all the tests in a workplace to fix this.

Reply:
@Emilgardis this would prevent all the tests from running in parallel (as-is), as they would be competing for a single workspace lock.

I also tried putting them in a single workspace and ran a `cargo build --all` at the workspace level, and that seemed to be possible theoretically, but that caused some spurious failures in my testing (perhaps resource exhaustion?). If you feel strongly about it, I can reinvestigate, though I would probably think we could address this as an improvement in the future.